crates/turborepo-lib/src/global_deps_package_detector.rs

```diff
@@ -0,0 +1,146 @@
+use thiserror::Error;
+use turbopath::AnchoredSystemPath;
+use turborepo_repository::{
+    change_mapper::{DefaultPackageDetector, PackageDetection, PackageDetector},
+    package_graph::{PackageGraph, WorkspacePackage},
+};
+use wax::{BuildError, Program};
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    InvalidFilter(#[from] BuildError),
+}
+
+/// A package detector that uses a global deps list to determine
+/// if a file should cause all packages to be marked as changed.
+/// This is less conservative than the `DefaultPackageDetector`
+/// which assumes that any changed file that is not in a package
+/// changes all packages. Since we have a list of global deps,
+/// we can check against that and avoid invalidating in unnecessary cases.
+pub struct GlobalDepsPackageDetector<'a> {
+    pkg_dep_graph: &'a PackageGraph,
+    global_deps_matcher: wax::Any<'a>,
+}
+
+impl<'a> GlobalDepsPackageDetector<'a> {
+    pub fn new<S: wax::Pattern<'a>, I: Iterator<Item = S>>(
+        pkg_dep_graph: &'a PackageGraph,
+        global_deps: I,
+    ) -> Result<Self, Error> {
+        let global_deps_matcher = wax::any(global_deps)?;
+
+        Ok(Self {
+            pkg_dep_graph,
+            global_deps_matcher,
+        })
+    }
+}
+
+impl<'a> PackageDetector for GlobalDepsPackageDetector<'a> {
+    fn detect_package(&self, path: &AnchoredSystemPath) -> PackageDetection {
+        match DefaultPackageDetector::new(self.pkg_dep_graph).detect_package(path) {
+            // Since `DefaultPackageDetector` is overly conservative, we can check here if
+            // the path is actually in globalDeps and if not, return it as
+            // PackageDetection::Package(WorkspacePackage::root()).
+            PackageDetection::All => {
+                let cleaned_path = path.clean();
+                let in_global_deps = self.global_deps_matcher.is_match(cleaned_path.as_str());
+
+                if in_global_deps {
+                    PackageDetection::All
+                } else {
+                    PackageDetection::Package(WorkspacePackage::root())
```

Comment:
So I think there's a few misunderstandings (not helped by the multiple iterations of this PR):

- It is helpful to know that the root package changed, because you can have root tasks that can get run, say if you define a `//#build` entry in pipeline. We can't remove this root package return because that would be a breaking change for people with root package tasks.
- Root package changed does *not* imply all packages changed. I thought it did but @chris-olszewski explained to me that it does not. I think it should in an ideal world (and I think a setting to enable that behavior is worth discussing but out of scope).
- Therefore separating between the two concepts is fine because that is the existing behavior.
- As for how our model relates to a package manager, I agree that we should stay closer to the package manager's semantics, but unfortunately we can't due to existing behavior. Like we should treat a root package change as a global change because that's what package managers effectively do, but we have to wait for a breaking change to make that possible.